controller/src/main/java/io/pravega/controller/server/bucket/BucketService.java

```diff
@@ -0,0 +1,283 @@
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This class represents an instance of a background worker service that runs background work for all streams under a 
+ * specific bucket for a specific service type. This means a separate instance of this class is instantiated for each 
+ * bucket for each service type.   
+ * Each object has two categories of work that it performs - process notification and execute the supplied worker funtion 
+ * on known sets of streams.
+ * 
+ * This is an abstract class and its implementations should monitor the underlying bucket store and call `notify` method 
+ * whenever streams are added to or removed from the bucket. These stream change notifications are added to the notification queue.
+ * 
+ * The notification loop periodically dequeues at most `MAX_NOTIFICATIONS_TO_TAKE` notifications from the queue and asynchronously 
+ * process these notifications. Upon completion of notification, the loop continues with the next iteration. 
+ * The queue used for notification is `BlockingDrainingQueue` which exposes an asynchronous `take` method. 
+ * 
+ * Notification loop scans the notification queue and if a new stream has been added to the bucket, it
+ * posts a new entry into work priority queue.  
+ * 
+ * Work posted into the priority queue has `stream identifier` and `time`. The queue is ordered based on time when the work 
+ * should be executed next. We ensure that there is exactly one work per stream that is posted into the work priority queue. 
+ * 
+ * Worker loop peeks at the first element from the priority queue. If the first element in the queue can be executed, 
+ * it is taken from the queue else, the loop sleeps till it can dequeue a work. 
+ * 
+ * Worker loop runs as an infinite loop and, at the most, dequeues `availableSlots` number of work items from the priority queue. 
+ * During each loop iteration it first checks if there are slots available to pick a work. If not, it postpones itself with a delay. 
+ * Available slots is managed via a thread safe counter. Whenever a new work is picked from the queue, the counter is decremented 
+ * and work is started asynchronously. 
+ * Whenever the work completes, we add the entry back into the work queue with next schedule for the stream after `execution duration` 
+ * and increment the available slots counter so that more outstanding work from the queue can be picked up. 
+ * This ensures that we only have a limited number of outstanding work items irrespective of number of streams under the bucket. 
+ * With exactly one entry per stream in the priority queue, we also ensure fairness. 
+ */
+abstract class BucketService extends AbstractService {
```

Comment:
It only keeps the stream id in memory (Scope and stream names) for all streams. 
All processing related objects are created and queued up only for "max-concurrent-execution" number of streams. 

Lets say scope and stream names are 50 bytes each, it would mean 100 bytes per stream. 
So if buckets owned by current controller has say 1M streams, it would mean 1M * 100 = 100 Mb of additional memory footprint.